modules/promotion/src/Element/CouponRedemptionForm.php

```diff
@@ -0,0 +1,225 @@
+    /** @var \Drupal\commerce_order\Entity\OrderInterface $order */
+    $order = $form_state->get('order');
+    $entity_type_manager = \Drupal::entityTypeManager();
+
+    // Validate that the coupon code exists.
+    $coupon = static::loadCouponByCode($coupon_code);
+    if (empty($coupon)) {
+      $form_state->setErrorByName($code_path, t('Coupon is invalid'));
+      return;
+    }
+
+    foreach ($order->get('coupons') as $item) {
+      if ($item->target_id == $coupon->id()) {
+        $form_state->setErrorByName($code_path, t('Coupon is already redeemed'));
+        return;
+      }
+    }
+
+    // If it does, ensure its promotion is still valid.
+    $order_type_storage = $entity_type_manager->getStorage('commerce_order_type');
+    /** @var \Drupal\commerce_promotion\PromotionStorageInterface $promotion_storage */
+    $promotion_storage = $entity_type_manager->getStorage('commerce_promotion');
+
+    /** @var \Drupal\commerce_order\Entity\OrderTypeInterface $order_type */
+    $order_type = $order_type_storage->load($order->bundle());
+    $promotion = $promotion_storage->loadByCoupon($order_type, $order->getStore(), $coupon);
+    if (empty($promotion)) {
+      $form_state->setErrorByName($code_path, t('Coupon is invalid'));
+      return;
+    }
+
+    $promotion_applies = FALSE;
+    if ($promotion->applies($order)) {
+      $promotion_applies = TRUE;
+    }
+    else {
+      foreach ($order->getItems() as $orderItem) {
+        if ($promotion->applies($orderItem)) {
+          $promotion_applies = TRUE;
+          break;
+        }
+      }
+    }
+
+    if ($promotion_applies === FALSE) {
+      $form_state->setErrorByName($code_path, t('Coupon is invalid'));
+      return;
+    }
+  }
+
+  /**
+   * Submits the coupon redemption form.
+   *
+   * @param array $element
+   *   The form element.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   */
+  public static function submitForm(array &$element, FormStateInterface $form_state) {
+    $coupon_parents = array_merge($element['#parents'], ['code']);
+    $coupon_code = $form_state->getValue($coupon_parents);
+    /** @var \Drupal\commerce_order\Entity\OrderInterface $order */
+    $order = $form_state->get('order');
+    $entity_type_manager = \Drupal::entityTypeManager();
+    $order_type_storage = $entity_type_manager->getStorage('commerce_order_type');
+    /** @var \Drupal\commerce_promotion\PromotionStorageInterface $promotion_storage */
+    $promotion_storage = $entity_type_manager->getStorage('commerce_promotion');
+
+    $coupon = static::loadCouponByCode($coupon_code);
+    /** @var \Drupal\commerce_order\Entity\OrderTypeInterface $order_type */
+    $order_type = $order_type_storage->load($order->bundle());
+    $promotion = $promotion_storage->loadByCoupon($order_type, $order->getStore(), $coupon);
+
+    if ($promotion->applies($order)) {
+      $order->get('coupons')->appendItem($coupon);
```

Comment:
Looks like we're missing a helper, since we're repeating the applies logic in two places. We could at least have one on the form element (then think of where to move it in a followup)